### ISSUE TITLE
Union with type that has optional field causes segfault in compiler

### ISSUE THREAD

**Issue by DEC05EBA** [ziglang/zig#1995]
If an union has a type with an optional type then the compiler with crash with segfault.
Example code (stripped down from larger project):
```zig
const LhsExpr = struct {
    rhsExpr: ?AstObject
};

const AstObject = union {
    lhsExpr: LhsExpr
};

test "crash" {
    const lhsExpr = LhsExpr {
        .rhsExpr = null
    };

    const obj = AstObject {
        .lhsExpr = lhsExpr
    };
}
```
The compiler crashes because it fails to allocate at src/analyze.cpp:2293 (resolve_union_type) because union_type->data.unionation.gen_field_count is 0

**Comment by andrewrk** [comment]
Thanks for the stripped down example.

**Comment by DEC05EBA** [comment]
I see now that it's not supposed to work because there is a recursive dependency (LhsExpr depends on a non-pointer type of AstObject and AstObject depends on a non-pointer type of LhsExpr).
So this works and doesn't crash:
```zig
const std = @import("std");

pub const LhsExpr = struct {
    rhsExpr: ?AstObject
};

pub const AstObject = union {
    lhsExpr: *LhsExpr
};

test "crash" {
    var lhsExpr = try std.debug.global_allocator.create(LhsExpr);
    lhsExpr.* = LhsExpr {
        .rhsExpr = null
    };

    const astObj = AstObject {
        .lhsExpr = lhsExpr
    };
}
```

however this does crash:
```zig
const std = @import("std");

pub const LhsExpr = struct {
    rhsExpr: ?AstObject
};

pub const AstObject = union {
    lhsExpr: *LhsExpr
};

test "crash" {
    const alignment = @alignOf(AstObject);
}
```
on the other hand `@alignOf(LhsExpr)` doesn't crash

**Comment by daurnimator** [comment]
I think this might be a dupe of #1735

**Comment by DEC05EBA** [comment]
Might be for the first example however for the last example it seems to be different as im using a pointer so it's not an issue of recursive dependency

**Comment by DEC05EBA** [comment]
#1735 is different than the last example as it's not supposed to compile since there is a recursive dependency but the compiler should give an error instead of crashing, while recursive dependency shouldn't matter for pointers as pointer size doesn't change depending on the size of the struct. I'll try to see if i can fix both issues

**Comment by andrewrk** [comment]
Proposal to fix this: #2174
